Refactor e2e tests

mahalrs opened the pull request:
Part of #315

This PR refactors e2e tests that will make it easier to track issues when running tests in parallel (with `-p` flag).

Commit:
[WIP] Parallelize e2e tests

Commit:
Fix golangci-lint

Commit:
[WIP] Parallelize e2e tests

Commit:
Fix golangci-lint

Commit:
Add test cases to run in parallel across 4 nodes

Commit:
Refactor test code

Commit:
Refactor test utils

Commit:
Minor changes to tests

Commit:
Merge branch 'main' into parallel-tests

nitishm commented:
@mahalrs It still seems like it is taking a long time - 28 minutes on the last one https://github.com/Azure/orkestra/pull/343/checks?check_run_id=2923294379
Do you know what the bottleneck might be?

mahalrs commented:
> @mahalrs It still seems like it is taking a long time - 28 minutes on the last one https://github.com/Azure/orkestra/pull/343/checks?check_run_id=2923294379
> Do you know what the bottleneck might be?

I disabled parallel tests. If I run them in parallel, tests are failing randomly. I still need to find the issue. Since it is the first test that's failing (as mentioned in other PR), we can merge this (can change this PR to a code refactor PR and will make another PR once I figure out the issue with running tests in parallel).

Commit:
Update Makefile

Commit:
Update Makefile

nitishm commented:
@jonathan-innis can you eye-ball this and see if it looks good? We can get it merged to unblock #340

nitishm commented on controllers/appgroup_controller_suite_test.go:
```diff
@@ -34,49 +31,68 @@ import (
+	tempChartStoreTargetDir string
+	testEnv                 *envtest.Environment
+
+	AfterEach  = ginkgo.AfterEach
+	BeforeEach = ginkgo.BeforeEach
+	By         = ginkgo.By
+	Context    = ginkgo.Context
+	Describe   = ginkgo.Describe
+	It         = ginkgo.It
+
+	BeNil        = gomega.BeNil
+	BeTrue       = gomega.BeTrue
+	Equal        = gomega.Equal
+	Eventually   = gomega.Eventually
+	Expect       = gomega.Expect
+	HaveOccurred = gomega.HaveOccurred
```
This might cause confusion for newer devs coming from a ginkgo background. Let's keep ginkgo and gomega as a `.` import for consistency, unless you have a good reason to do what you are doing here?

nitishm commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,430 @@
+		key      types.NamespacedName
+	)
+
+	BeforeEach(func() {
```
Do we no longer need to create the namespace?

nitishm commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,430 @@
+		By("Checking that the all the HelmReleases have come up and are in a ready state")
+		err = k8sClient.List(ctx, hrList, client.InNamespace(name))
+		Expect(err).NotTo(HaveOccurred())
+		Expect(len(hrList.Items)).To(Equal(oldHelmReleaseCount + OnlyApplicationsHelmReleaseCount))
```
s/OnlyApplicationsHelmReleaseCount/onlyApplicationsHelmReleaseCount/
Doesn't need to be exported.

nitishm commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,430 @@
-
-		})
-
-		It("Should create create only application releases with subchart nil successfully", func() {
-			applicationGroup := defaultAppGroup(name)
-			applicationGroup.Name = name
-			applicationGroup.Namespace = DefaultNamespace
-			for i := range applicationGroup.Spec.Applications {
-				applicationGroup.Spec.Applications[i].Spec.Subcharts = nil
+		}
```
This seems to be used across test cases. Maybe create a helper test function for this?

mahalrs replied:
Fixed

nitishm commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,430 @@
+		}, DefaultTimeout, time.Second).Should(BeTrue())
+	})
+
+	It("Should create the bookinfo and then delete it while in progress", func() {
```
Is there a similar test for update while in progress?

mahalrs replied:
there isn't

nitishm commented on controllers/appgroup_controller_testutils_test.go:
```diff
@@ -1,15 +1,15 @@
-package controllers
+package controllers_test
```
The file name is loooong. Any particular reason for switching it from testutils?

mahalrs replied:
I changed package to controllers_test. So, I need to change testutils.go to testutils_test.go. This might confuse someone to think that testutils_test.go contains tests for testutils.go. If you want, I can change this to testutils_test.go

nitishm commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,430 @@
 
 var _ = Describe("ApplicationGroup Controller", func() {
+	const (
+		DefaultNamespace                 = "orkestra"
+		DefaultTimeout                   = time.Minute * 5
+		TotalHelmReleaseCount            = 6
+		OnlyApplicationsHelmReleaseCount = 2
```
None of there need to be uppercase.

nitishm reviewed (changes requested):
NITs and Questions

jonathan-innis commented:
I don't see the point of most of these changes. @mahalrs if you could help me understand why these changes were made, that would be great

Commit:
Unexport consts, add helper functions

mahalrs commented:
> I don't see the point of most of these changes. @mahalrs if you could help me understand why these changes were made, that would be great

Did some cleanup to make it easier to track issues related to parallel tests. Some changes made are:

- moved cleanup code to AfterEach
- moved initializing default application group to BeforeEach
- moved creating a key object to BeforeEach
- added some helper functions to use across test cases
- some minor changes for readability

jonathan-innis commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,419 @@
+
+	var (
+		ctx context.Context
+		err error
```
Why do we need errors at this scope?

mahalrs replied:
Fixed

jonathan-innis commented on controllers/testutils_test.go:
```diff
@@ -29,31 +31,82 @@ const (
 )
 
-func defaultAppGroup(targetNamespace string) *v1alpha1.ApplicationGroup {
+func objKeyBuilder(name, namespace string) types.NamespacedName {
```
I think this function can be removed as someone can just right `types.NamespacedName{Name:...., Namespace:...}` and it accomplishes the same thing. Adding a function for this is redundant

mahalrs replied:
Fixed

jonathan-innis commented on controllers/testutils_test.go:
```diff
@@ -29,31 +31,82 @@ const (
+	}
+}
+
+func workflowBuilder(name, namespace string) *v1alpha13.Workflow {
```
This also abstracts the functionality away for no reason in a test case. IMO, this makes it more difficult to read and understand what the test case is accomplishing

mahalrs replied:
Fixed

jonathan-innis commented on controllers/testutils_test.go:
```diff
@@ -1,15 +1,17 @@
-package controllers
+package controllers_test
```
I like the package changes; however, the name of the file isn't great. Maybe `utils_test`?

jonathan-innis commented on controllers/testutils_test.go:
```diff
@@ -29,31 +31,82 @@ const (
 }
 
-func smallAppGroup(targetNamespace string) *v1alpha1.ApplicationGroup {
+func smallAppGroup(groupName, groupNamespace, targetNamespace string) *v1alpha1.ApplicationGroup {
```
`name` and `namespace` are reasonable names IMO, no need to change this

jonathan-innis commented on controllers/testutils_test.go:
```diff
@@ -29,31 +31,82 @@ const (
+	return appGroup
+}
+
+func defaultAppGroup(groupName, groupNamespace, targetNamespace string) *v1alpha1.ApplicationGroup {
```
`name` and `namespace` are reasonable names IMO, no need to change this

Commit:
Remove redundant helper functions, rename testutils

nitishm reviewed (dismissed):
🚀  LGTM

mahalrs commented on controllers/appgroup_controller_test.go:
```diff
@@ -1,556 +1,420 @@
+		defaultNamespace                 = "orkestra"
+		defaultTimeout                   = time.Minute * 5
+		totalHelmReleaseCount            = 6
+		onlyApplicationsHelmReleaseCount = 2
```
It's the number of application helm releases (ambassador & bookinfo), excluding subcharts.

jonathan-innis replied:
Can you create comments next to these variables so that both `totalHelmReleaseCount` and `onlyApplicationsHelmReleaseCount` is clear

Commit:
Add comments about helm release count variables

jonathan-innis reviewed (approved):
🚀
